internal/wire/wire.go

```diff
@@ -257,13 +258,16 @@ func newGen(pkg *packages.Package) *gen {
 }
 
 // frame bakes the built up source body into an unformatted Go source file.
-func (g *gen) frame() []byte {
+func (g *gen) frame(tags string) []byte {
 	if g.buf.Len() == 0 {
 		return nil
 	}
 	var buf bytes.Buffer
+	if len(tags) > 0 {
+		tags = fmt.Sprintf(" gen -tags \"%s\"", tags)
+	}
 	buf.WriteString("// Code generated by Wire. DO NOT EDIT.\n\n")
-	buf.WriteString("//go:generate wire\n")
+	buf.WriteString("//go:generate wire" + tags + "\n")
```

Comment:
This is the part that concerns me: I'm pretty firmly convinced that if the generated code should include additional build tags, that this should be part of the template file, not a command-line flag. The key problem here (as well as in #48) is that we don't have a good way of generating a file for each configuration (i.e. test, not-test, and now build tag combinations). My understanding was that this was purely about having analysis complete, not about changing the generated output.

Reply:
In my case, if I'm targeting device type A, that is my build target (specified by build tags) until I move to a different issue. I don't want that target to change until I explicitly change the build configuration to device "B" (by rebuilding and re-wireing with -tags "B").

So are you saying we should list all possible build tags in wire.go, then generate a file for each combination? We could potentially generate a lot of files that way.

Or another way would be to have separate wire.go file for each build tags, but I could have done that already without this patch :-) This way would defeat the point of code generation.

Reply:
Yes, I am saying that Wire should be generating a distinct file for each combination of build tags, and each distinct template with a different build tag configuration should be in a separate file.

I am having difficulty understanding what your use case is. It seems like you are trying to generate multiple different _templates_ that are in different build configurations. I don't think that's a use case we want to support in Wire.

Reply:
The following [testscript](https://github.com/rogpeppe/go-internal/tree/master/testscript) is a basic repro of my test case:

```
go run -tags=deviceA .
stdout something

go run -tags=deviceB .
stdout 42

-- device/driver.go --
package device

type Driver interface {
        NextID() DeviceID
}

-- device/impl-a.go --
//+build deviceA

package device

type DeviceID string

type DeviceA struct {
}

func (d *DeviceA) NextID() DeviceID {
        return "something"
}

func NewDriver() Driver {
        return &DeviceA{}
}

-- device/impl-b.go --
//+build deviceB

package device

type DeviceID int

type DeviceB struct {
}

func (d *DeviceB) NextID() DeviceID {
        return 42
}

func NewDriver() Driver {
        return &DeviceB{}
}

-- main.go --
package main

import (
        "./device"
        "fmt"
)

func main() {
        driver := device.NewDriver()
        fmt.Println(driver.NextID())
}
```

Creating a driver in my real-world use case is more complicated, so I want to use wire to inject those dependencies. To do this, I will need to pass build tags to the code analyser.

```
# setup wire & deps...
go get github.com/mwlazlo/wire
go get github.com/google/wire # download runtime library as well, since package path is hardcoded into wire...
cd $WORK/gopath/src/github.com/mwlazlo/wire/
go mod download
go install ./cmd/wire
cd $WORK
exists $WORK/gopath/bin/wire

# work on deviceA
exec $WORK/gopath/bin/wire gen -tags=deviceA ./device/
exists device/wire_gen.go
go run -tags=deviceA .
stdout something

# first case works, try re-generating code with go generate

go generate ./device
go run -tags=deviceA .
stdout something

# still works after running go generate

# work on deviceB
exec $WORK/gopath/bin/wire gen -tags=deviceB ./device/
exists device/wire_gen.go
go run -tags=deviceB .
stdout 42

-- device/driver.go --
package device

type Driver interface {
        NextID() DeviceID
}

-- device/impl-a.go --
//+build deviceA

package device

type DeviceID string

type DeviceA struct {
}

func (d *DeviceA) NextID() DeviceID {
        return "something"
}

func NewDriver() Driver {
        return &DeviceA{}
}

-- device/impl-b.go --
//+build deviceB

package device

type DeviceID int

type DeviceB struct {
}

func (d *DeviceB) NextID() DeviceID {
        return 42
}

func NewDriver() Driver {
        return &DeviceB{}
}

-- main.go --
package main

import (
        "./device"
        "fmt"
)

func main() {
        driver := device.ProvideDriver()
        fmt.Println(driver.NextID())
}

-- device/wire.go --
//+build wireinject

package device

import "github.com/google/wire"

func ProvideDriver() Driver {
        panic(wire.Build(NewDriver))
}
```

Reply:
I've thought about this more this last week, and while I do think this deserves a longer-term strategy so that an invocation per target configuration is not necessary, I don't think that should block this PR. Sorry for the back and forth; your explanation helped me understand the problem better.

Reply:
Not at all, it's good to be clear before accepting changes. I've updated the PR.